### PR TITLE
skiplist: switch to alloc API

### PR DESCRIPTION
**Pull request by tobz**
This addresses #246 by removing the allocation hack that was in place before.

Signed-off-by: Toby Lawrence <toby@nuclearfurnace.com>

### CODE REVIEW

**Comment by tobz** [comment]
@stjepang I think this is all set now.

**Comment by tobz** [comment]
Hmm, not really sure why it thinks that `std::alloc` doesn't exist... that's straight out of the Rust docs themselves. O_o

**Comment by stjepang** [comment]
It's because of [these shenanigans](https://github.com/crossbeam-rs/crossbeam/blob/711fd376b99aefaea9a3167033bd56a354d87dc2/crossbeam-skiplist/src/lib.rs#L13-L22) we use to support no_std + alloc environments. You probably need `use alloc::alloc` instead of `use std::alloc`.

**Comment by tobz** [comment]
Ahhhh, alright.  Lemme whip up the fix.

**Comment by tobz** [comment]
Looks like it's an unrelated `crossbeam-channel` failure at this point?

**Comment by stjepang** [comment]
That is an unrelated regression on nightly Rust: https://github.com/rust-lang/rust/issues/56867

File: crossbeam-skiplist/src/base.rs [hunk 1]
```diff
@@ -99,11 +99,8 @@ impl<K, V> Node<K, V> {
     /// with null pointers. However, the key and the value will be left uninitialized, and that is
     /// why this function is unsafe.
     unsafe fn alloc(height: usize, ref_count: usize) -> *mut Self {
-        // TODO(stjepang): Use the new alloc API instead of this hack once it becomes stable.
-        let cap = Self::size_in_u64s(height);
-        let mut v = Vec::<u64>::with_capacity(cap);
-        let ptr = v.as_mut_ptr() as *mut Self;
-        mem::forget(v);
+        let layout = Self::get_layout(height);
+        let ptr = alloc(layout) as *mut Self;
```
Comment (SimonSapin) [inline crossbeam-skiplist/src/base.rs]: This needs to check for null and call `handle_alloc_error(layout)`.

Comment (stjepang) [reply crossbeam-skiplist/src/base.rs]: Oh, thanks for pointing this out! I'm used to `Box::new()` always working so the possibility of a failure didn't even cross my mind.

@tobz Would you like to submit a fix for this?

Comment (tobz) [reply crossbeam-skiplist/src/base.rs]: I'll submit a fix, yeah.

**Pull request merged**